openedx/core/djangoapps/course_groups/models.py

```diff
@@ -92,52 +87,47 @@ def clean(self):
-        with outer_atomic(read_committed=True):
-
-            saved_membership, created = CohortMembership.objects.select_for_update().get_or_create(
-                user__id=self.user.id,
-                course_id=self.course_id,
+    @classmethod
+    def assign(cls, cohort, user):
+        """
+        Assign user to cohort, switching them to this cohort if they had previously been assigned to another
+        cohort
+        Returns CohortMembership, previous_cohort (if any)
+        """
+        with transaction.atomic():
+            membership, created = cls.objects.select_for_update().get_or_create(
+                user__id=user.id,
+                course_id=cohort.course_id,
                 defaults={
-                    'course_user_group': self.course_user_group,
-                    'user': self.user
-                }
-            )
+                    'course_user_group': cohort,
+                    'user': user
+                })
 
-            # If the membership was newly created, all the validation and course_user_group logic was settled
-            # with a call to self.save(force_insert=True), which gets handled above.
             if created:
-                return
-
-            if saved_membership.course_user_group == self.course_user_group:
+                membership.course_user_group.users.add(user)
+                previous_cohort = None
+            elif membership.course_user_group == cohort:
                 raise ValueError("User {user_name} already present in cohort {cohort_name}".format(
-                    user_name=self.user.username,
-                    cohort_name=self.course_user_group.name
-                ))
-            self.previous_cohort = saved_membership.course_user_group
-            self.previous_cohort_name = saved_membership.course_user_group.name
-            self.previous_cohort_id = saved_membership.course_user_group.id
-            self.previous_cohort.users.remove(self.user)
+                    user_name=user.username,
+                    cohort_name=cohort.name))
+            else:
+                previous_cohort = membership.course_user_group
+                previous_cohort.users.remove(user)
+
+                membership.course_user_group = cohort
+                membership.course_user_group.users.add(user)
+                membership.save()
+        return membership, previous_cohort
+
+    def save(self, force_insert=False, force_update=False, using=None, update_fields=None):
+        self.full_clean(validate_unique=False)
 
-            saved_membership.course_user_group = self.course_user_group
-            self.course_user_group.users.add(self.user)
+        log.info("Saving CohortMembership for user '%s' in '%s'", self.user.id, self.course_id)
 
-            super(CohortMembership, saved_membership).save(update_fields=['course_user_group'])
+        return super(CohortMembership, self).save(force_insert=force_insert,
```

Comment:
style nit: we prefer to break all of the arguments in long calls to newlines, so that these aren't all indented so far (admittedly, we are not always great about sticking to the style guide).

https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#breaking-long-lines

Reply:
pep8 complained unless I indented it exactly as shown. If that's the case, the style guide conflicts with the quality tests.

Reply:
Oh wow, that's weird.  The vast majority of our indentation adheres to the style guide.  If the quality test is passing, it's probably not worth changing.  Maybe we can take run pep8 locally tomorrow and see what's up, just in case the pep8 config went whacky for a minute.